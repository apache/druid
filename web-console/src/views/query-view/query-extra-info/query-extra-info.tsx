/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  Button,
  Intent,
  Menu,
  MenuDivider,
  MenuItem,
  Popover,
  Position,
  Tooltip,
} from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import copy from 'copy-to-clipboard';
import { QueryResult } from 'druid-query-toolkit';
import React from 'react';

import { AppToaster } from '../../../singletons';
import { pluralIfNeeded } from '../../../utils';

import './query-extra-info.scss';

export interface QueryExtraInfoProps {
  queryResult: QueryResult;
  onDownload: (filename: string, format: string) => void;
}

export const QueryExtraInfo = React.memo(function QueryExtraInfo(props: QueryExtraInfoProps) {
  const { queryResult, onDownload } = props;

  function handleQueryInfoClick() {
    const id = queryResult.queryId || queryResult.sqlQueryId;
    if (!id) return;

    copy(id, { format: 'text/plain' });
    AppToaster.show({
      message: 'Query ID copied to clipboard',
      intent: Intent.SUCCESS,
    });
  }

  function handleDownload(format: string) {
    const id = queryResult.queryId || queryResult.sqlQueryId;
    if (!id) return;

    onDownload(`query-${id}.${format}`, format);
  }

  const downloadMenu = (
    <Menu className="download-format-menu">
      <MenuDivider title="Download as:" />
      <MenuItem text="CSV" onClick={() => handleDownload('csv')} />
      <MenuItem text="TSV" onClick={() => handleDownload('tsv')} />
      <MenuItem text="JSON (new line delimited)" onClick={() => handleDownload('json')} />
    </Menu>
  );

  const wrapQueryLimit = queryResult.getSqlOuterLimit();
  let resultCount: string;
  if (wrapQueryLimit && queryResult.getNumResults() === wrapQueryLimit) {
    resultCount = `${queryResult.getNumResults() - 1}+ results`;
  } else {
    resultCount = pluralIfNeeded(queryResult.getNumResults(), 'result');
  }

  let tooltipContent: JSX.Element | undefined;
  if (queryResult.queryId) {
    tooltipContent = (
      <>
        Query ID: <strong>{queryResult.queryId}</strong> (click to copy)
      </>
    );
  } else if (queryResult.sqlQueryId) {
    tooltipContent = (
      <>
        SQL query ID: <strong>{queryResult.sqlQueryId}</strong> (click to copy)
      </>
    );
  }

  return (
    <div className="query-extra-info">
      {typeof queryResult.queryDuration !== 'undefined' && (
        <div className="query-info" onClick={handleQueryInfoClick}>
          <Tooltip content={tooltipContent} hoverOpenDelay={500}>
            {`${resultCount} in ${(queryResult.queryDuration / 1000).toFixed(2)}s`}
          </Tooltip>
        </div>
      )}
      <Popover className="download-button" content={downloadMenu} position={Position.BOTTOM_RIGHT}>
        <Button icon={IconNames.DOWNLOAD} minimal />
      </Popover>
    </div>
  );
});
