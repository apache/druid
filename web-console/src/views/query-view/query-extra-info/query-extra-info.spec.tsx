/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { render } from '@testing-library/react';
import React from 'react';

import { QueryExtraInfo } from './query-extra-info';

describe('query extra info', () => {
  it('matches snapshot', () => {
    const queryExtraInfo = (
      <QueryExtraInfo
        queryExtraInfo={{
          queryId: 'e3ee781b-c0b6-4385-9d99-a8a1994bebac',
          startTime: new Date('1986-04-26T01:23:40+03:00'),
          endTime: new Date('1986-04-26T01:23:48+03:00'),
          numResults: 2000,
          wrappedLimit: 2000,
        }}
        onDownload={() => {}}
      />
    );

    const { container } = render(queryExtraInfo);
    expect(container.firstChild).toMatchSnapshot();
  });
});
