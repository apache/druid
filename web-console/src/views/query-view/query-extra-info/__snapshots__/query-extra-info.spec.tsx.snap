// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query extra info matches snapshot 1`] = `
<div
  class="query-extra-info"
>
  <div
    class="query-info"
  >
    <span
      class="bp3-popover-wrapper"
    >
      <span
        class="bp3-popover-target"
      >
        <span
          class=""
          tabindex="0"
        >
          999+ results in 8.00s
        </span>
      </span>
    </span>
  </div>
  <span
    class="bp3-popover-wrapper download-button"
  >
    <span
      class="bp3-popover-target"
    >
      <button
        class="bp3-button bp3-minimal"
        type="button"
      >
        <span
          class="bp3-icon bp3-icon-download"
          icon="download"
        >
          <svg
            data-icon="download"
            height="16"
            viewBox="0 0 16 16"
            width="16"
          >
            <desc>
              download
            </desc>
            <path
              d="M7.99-.01c-4.42 0-8 3.58-8 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zM11.7 9.7l-3 3c-.18.18-.43.29-.71.29s-.53-.11-.71-.29l-3-3A1.003 1.003 0 015.7 8.28l1.29 1.29V3.99c0-.55.45-1 1-1s1 .45 1 1v5.59l1.29-1.29a1.003 1.003 0 011.71.71c0 .27-.11.52-.29.7z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </button>
    </span>
  </span>
</div>
`;
